### ISSUE TITLE
Support using stored procedures in a Database

### ISSUE THREAD

**Issue by agilliland** [metabase/metabase#1532]
It's possible for a database to describe a query that a user might run as a stored procedure and thus want to allow that stored procedure to be available in Metabase for users to access somehow.

Here's an example provided by @camsaul 

```sql
CREATE PROCEDURE username_for_id

@user_id int

AS

SELECT "username" FROM "users"
WHERE user_id = @user_id
```

and then:

```sql
exec username_for_id '1'
```

**Comment by mattkrea** [comment]
In our case we do often then run queries on top of this data. Say, for example, we have a complicated stored procedure to calculate processing volume and profit for a large customer base (we are in payment processing so this comes to mind). This SP can be sent off for review on it's own or further manipulated especially considering some of these have hundreds of thousands of records or multiple result sets returned.

**Comment by billsinc** [comment]
Yes, I don't know if SP are "queryable" and would then populate a list or if we would have to specify them manually in some sort of text entry field.

**Comment by AlgyTaylor** [comment]
:+1: from me, quite often use stored procedures & functions to reduce a bit of cognitive workload

**Comment by coussej** [comment]
It would indeed be very usefull to provide an option to select a stored procedure, fill in the input parameters, and allow users to play around with the result set.

**Comment by dacort** [comment]
:+1: Use case would be creating a generic dashboard and then allowing user to select input parameters via manual input or other tables.

**Comment by paoliniluis** [comment]
update: if you use functions instead of SP's, it works as a charm.

e.g. (using the sample dataset):
```
CREATE function get_all_people_from_state(s char(2))
    RETURNS TABLE(
    address varchar(255),
    email varchar(255),
    name varchar(255)
)
LANGUAGE SQL
AS $$
SELECT address, email, name FROM PEOPLE WHERE state = s
$$;
```
and then do:
```
SELECT * from get_all_people_from_state({{state}})
```

you can hide all query logic in prepared statements and use a function from metabase native sql query editor

**Comment by krispharper** [comment]
@paoliniluis Does this still work for you? When I try it I get a message that says, "Variables in models aren't supported yet."

For the broader group, this is something we'd really love to have. We have quite a bit of code in parameterized Postgres functions. We'd like to allow people to interact with those via Metabase by putting in parameters in the UI.

**Comment by paoliniluis** [comment]
@krispharper are you using a model or a question?

**Comment by poorbadger** [comment]
We would love this as well. We'd rather have our queries encapsulated and reusable within our databases, versioning and source control of data access logic, etc.
